vyper/codegen/core.py

```diff
@@ -982,7 +979,14 @@ def make_setter(left, right, hi=None):
     # Complex Types
     assert isinstance(left.typ, (SArrayT, TupleT, StructT))
 
-    return _complex_make_setter(left, right, hi=hi)
+    ret = ["seq"]
+    if hi is not None:
+        item_end = add_ofst(right, right.typ.abi_type.static_size())
+        len_check = ["assert", ["le", item_end, hi]]
+        ret.append(len_check)
+
+    ret.append(_complex_make_setter(left, right, hi=hi))
+    return IRnode.from_list(ret)
```

Comment:
`make_setter`'s caller always takes care of it from what I saw, but it wouldn't hurt to `cache_when_complex` `right` as it is done in the other cases.

Reply:
https://github.com/vyperlang/vyper/pull/4091/commits/a10662ba3b1d61fc257e3bf917842587eb7e4e42